sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+        prior: Prior distribution that is the same for all posteriors. If it is not the
+            same, then `prior = None`.
+        device: Device which the component distributions sit on.
+        default_x: Used in `.sample(), .log_prob()` as default conditioning context.
+    """
+
+    def __init__(
+        self,
+        posteriors: List,
+        weights: Optional[Union[List[float], Tensor]] = None,
+        theta_transform: Optional[TorchTransform] = None,
+    ):
+        """
+        Args:
+            posteriors: List containing the trained posterior instances that will make
+                up the ensemble.
+            weights: Assign weights to posteriors manually, otherwise they will be
+                weighted with 1/N.
+            theta_transform: If passed, this transformation will be applied during the
+            optimization.
+        """
+        self.posteriors = posteriors
+        self.num_components = len(posteriors)
+        self.weights = weights
+        self.potential_fn = None  # will be set when context is provided.
+
+        self.device = self.ensure_same_device(posteriors)
+
+        self.prior = self.common_prior()
+        if theta_transform is not None:
+            self.theta_transform = theta_transform
+        elif self.prior is not None:
+            self.theta_transform = mcmc_transform(self.prior, device=self.device)
+        else:
+            self.theta_transform = None
+
+    def common_prior(self) -> bool:
+        """Ensures that all posteriors in the ensemble are based off of the same prior
+        distribution.
+
+        Args:
+            posteriors: List containing the trained posterior instances that will make
+                up the ensemble.
+
+        Raises:
+            AssertionError if ensemble components have different priors.
+
+        Returns:
+            A prior distribution, if it is the same for all posteriors.
+        """
+        try:
+            # for SNPE
+            priors = [posterior.prior for posterior in self.posteriors]
+        except AttributeError:
+            # for SNLE and SNRE
+            priors = [posterior.proposal for posterior in self.posteriors]
+
+        # assert same type
+        same_type = all(isinstance(prior, type(priors[0])) for prior in priors)
+
+        def compare_params(dist_x, dist_y):
+            all_equal = True
+            for x, y in zip(dist_x.__dict__.values(), dist_y.__dict__.values()):
+                if type(x) == Tensor:
+                    all_equal = all_equal and torch.equal(x, y)
+                else:
+                    all_equal = all_equal and (x == y)
+            return all_equal
+
+        if same_type:
+            # assert same parameters
+            same_params = all(compare_params(prior, priors[0]) for prior in priors)
+
+        same_prior = same_type and same_params
+        if same_prior:
+            return priors[0]
+        else:
+            return None
```

Comment:
What happens if this is `None`? Will an error occur? If so, please raise here.

Reply:
I thought it might be desirable to keep the ensemble usable for posteriors with different proposals/priors. Therefore, I return the ensemble prior as None, if the components do not have the same prior. Prior is technically not needed anyway since the sampling and evaluations are done on each component independently. This is also the reason why I did not call the function ensure_same_prior, as I do not enforce it.

Reply:
So, if I got it right: the `common_prior` is only used to build the `theta_transform` and at `__init__()` of the EnsemblePosterior. However, it is only used in the `theta_transform` if the `theta_transform` is not passed directly. In addition, it is never used in the `EnsemblePosterior`.

I find it very confusing to be honest and I think that the code be largely simplified by getting rid of the `self.prior` attribute here (and do not do the check anymore).